harness/determined/_training.py

```diff
@@ -0,0 +1,167 @@
+        metrics: Dict[str, Any],
+        *,
+        total_records: Optional[int] = None,
+        total_epochs: Optional[int] = None,
+        computed_records: Optional[int] = None,
+    ) -> None:
+        serializable_metrics = set()
+        non_serializable_metrics = set()
+        # NaN and bytes are not JSON serializable.  In the case of trial implementation bugs or
+        # numerical instability issues, validation metric functions may return None or NaN values.
+        # For now, immediately fail any trial that encounters such a None metric. For NaN metrics,
+        # if it's the target of the searcher, we set it to +/- max_float depending on if the
+        # searcher is optimizing for the max or min. NaN metrics which are not the target of the
+        # searcher are dropped.
+        # TODO (DET-2495): Do not replace NaN metric values.
+        for metric_name, metric_value in metrics.items():
+            metric_is_none = metric_value is None
+            metric_is_nan = tensorboard.metric_writers.util.is_numerical_scalar(
+                metric_value
+            ) and math.isnan(metric_value)
+
+            if metric_is_none or metric_is_nan:
+                raise RuntimeError(
+                    "Validation metric '{}' returned "
+                    "an invalid scalar value: {}".format(metric_name, metric_value)
+                )
+
+            if isinstance(metric_value, (bytes, bytearray)):
+                non_serializable_metrics.add(metric_name)
+            else:
+                serializable_metrics.add(metric_name)
+
+        if len(non_serializable_metrics):
+            logger.warning(
+                "Removed non serializable metrics: %s", ", ".join(non_serializable_metrics)
+            )
+        reportable_metrics = {k: metrics[k] for k in serializable_metrics}
+
+        body = {
+            "trial_run_id": self._run_id,
+            "total_batches": total_batches,
+            "metrics": reportable_metrics,
+        }
+        if total_records is not None:
+            body["total_records"] = total_records
+        if total_epochs is not None:
+            body["total_epochs"] = total_epochs
+        if computed_records is not None:
+            body["computed_records"] = computed_records
+        logger.info(f"report_validation_metrics(total_batches={total_batches}, metrics={metrics})")
+        self._session.post(
+            f"/api/v1/trials/{self._trial_id}/validation_metrics",
+            data=det.util.json_encode(body),
+        )
+
+        # Also sync tensorboard (all metrics, not just json-serializable ones).
+        if self._tbd_writer and self._tbd_mgr:
+            self._tbd_writer.on_validation_step_end(total_batches, metrics)
+            self._tbd_mgr.sync()
+
+    def report_early_exit(self, reason: EarlyExitReason) -> None:
+        body = {"reason": EarlyExitReason(reason).value}
+        logger.info(f"report_early_exit({reason})")
+        self._session.post(
```

Comment:
probably should handle a 400 here or above in the call stack (which could be "you already told me this")